lib/translator/inquiry.cc

```diff
@@ -0,0 +1,289 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "inquiry.h"
+#include "string.h"
+
+namespace translator {
+
+// Creates and validates a Inquiry Command struct
+StatusCode RawToScsiCommand(absl::Span<const uint8_t> raw_cmd,
+                            scsi_defs::InquiryCommand &cmd) {
+  if (raw_cmd.empty()) {
+    DebugLog("buffer is empty or nullptr\n");
+    return StatusCode::kInvalidInput;
+  }
+
+  uint8_t raw_opcode = raw_cmd[0];
+  if (raw_opcode > 0xaf) {
+    DebugLog("raw opcode is out of range: %x", raw_opcode);
+    return StatusCode::kInvalidInput;
+  }
+
+  scsi_defs::OpCode opcode = static_cast<scsi_defs::OpCode>(raw_opcode);
+  if (!strcmp(ScsiOpcodeToString(opcode), "INVALID OPCODE")) {
+    DebugLog("invalid opcode - does not exist.");
+    return StatusCode::kInvalidInput;
+  }
+
+  if (opcode != scsi_defs::OpCode::kInquiry) {
+    const char *expected_cmd_str =
+        ScsiOpcodeToString(scsi_defs::OpCode::kInquiry);
+    const char *cmd_str = ScsiOpcodeToString(opcode);
+    DebugLog("wrong opcode. expected %s got %s.", expected_cmd_str, cmd_str);
+    return StatusCode::kInvalidInput;
+  }
+
+  // TODO: check invalid parameters in scsi_command
+  cmd = *reinterpret_cast<const scsi_defs::InquiryCommand *>(&raw_cmd[1]);
+  return StatusCode::kSuccess;
+}
+
+scsi_defs::InquiryData TranslateStandardInquiryResponse(
+    const nvme_defs::IdentifyControllerData &identify_controller_data,
+    const nvme_defs::IdentifyNamespace &identify_namespace_data) {
+  // SCSI Inquiry Standard Result
+  // https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+  // Section 6.1.1
+  scsi_defs::InquiryData result = scsi_defs::InquiryData{
+      .version = scsi_defs::Version::kSpc4,
+      .response_data_format = scsi_defs::ResponseDataFormat::kCompliant,
+      .additional_length = 0x1f,
+      .tpgs = scsi_defs::TPGS::kNotSupported,
+      .protect = (identify_namespace_data.dps.pit == 0 &&
+                  identify_namespace_data.dps.md_start == 0)
+                     ? 0
+                     : 1,
+
+      .cmdque = 1};
+
+  // Shall be set to “NVMe” followed by 4 spaces: “NVMe “
+  strncpy(result.vendor_identification, kNvmeVendorIdentification.data(), 8);
```

Comment:
It's worth noting that the vendor_identification string isn't null terminated.  While strncpy does the correct thing here and will leave a non-null-terminated string in result.vendor_identification, using memcpy might be more clear so that people are reminded it's arbitrary data rather than a legal C-string.  A comment (here or in the header) could also be helpful.